daisy/step_attach_disks.go

```diff
@@ -0,0 +1,130 @@
+//  Copyright 2017 Google Inc. All Rights Reserved.
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+package daisy
+
+import (
+	"context"
+	"fmt"
+	"path"
+	"sync"
+
+	"google.golang.org/api/compute/v1"
+)
+
+// AttachDisks is a Daisy AttachDisks workflow step.
+type AttachDisks []*AttachDisk
+
+// AttachDisk is used to attach a GCE disk ito an instance.
+type AttachDisk struct {
+	compute.AttachedDisk
+
+	// Instance to attach to.
+	Instance      string
+	project, zone string
+}
+
+func (a *AttachDisks) populate(ctx context.Context, s *Step) dErr {
+	for _, ad := range *a {
+		ad.Mode = strOr(ad.Mode, defaultDiskMode)
+		if ad.DeviceName == "" {
+			ad.DeviceName = path.Base(ad.Source)
+		}
+	}
+
+	return nil
+}
+
+func (a *AttachDisks) validate(ctx context.Context, s *Step) (errs dErr) {
+	for _, ad := range *a {
+		if !checkDiskMode(ad.Mode) {
+			fmt.Println("adding err")
+			errs = addErrs(errs, errf("cannot attach disk: bad disk mode: %q", ad.Mode))
+		}
+		if ad.Source == "" {
+			errs = addErrs(errs, errf("cannot attach disk: AttachedDisk.Source is empty"))
+		}
+
+		ir, err := s.w.instances.regUse(ad.Instance, s)
+		addErrs(errs, err)
+		if ir == nil {
+			// Return now, the rest of this function can't be run without ir.
+			return addErrs(errs, errf("cannot attach disk: instance %q not found in registry", ad.Instance))
+		}
+
+		if diskURLRgx.MatchString(ad.Source) {
+			ad.Source = extendPartialURL(ad.Source, ir.Project)
```

Comment:
This is a matter of opinion perhaps, but I've started moving the "if URL match, extend URL" logic into populate instead of validate.

If it can't be put in populate, could you put a comment here?